gcc/testsuite/rust/debug/win64-abi.rs

```diff
@@ -0,0 +1,11 @@
+pub extern "win64" fn square(num: i32) -> i32 {
+    num * num
+}
+
+fn main() {
+    // { dg-do compile { target { x86_64-*-* } } }
+    // { dg-options "-gdwarf-5 -dA -w -O1" }
+    // MS ABI dictates that the first argument is ecx instead of edi from the sysv world
+    // { dg-final { scan-assembler "%ecx, %ecx" } }
```

Comment:
Given a x86_64-pc-linux-gnu build, when running `make check-rust` with `RUNTESTFLAGS='--target_board=unix\{,-m32,-mx32\}'`, for example, we get `PASS` for the (default) `-m64` variant, but `FAIL: rust/debug/win64-abi.rs scan-assembler %ecx, %ecx` for the latter two variants.  The `target { x86_64-*-* }` does apply in this case (and in fact should be extended/generalized to `target i?86-*-* x86_64-*-*`), but will need separate `scan-assembler` directives, conditionalized with `target [...]` or `target { ! [...] }`, with `[...]` being `lp64`/`ia32`/`x32`/etc. as appropriate (see `gcc/doc/sourcebuild.texi`, "Effective-Target Keywords".

Reply:
Eh, actually for `-mx32`: `error: X32 does not support ‘ms_abi’ attribute`, so that will have to be handled differently, to expect that diagnostic instead.

Reply:
> Given a x86_64-pc-linux-gnu build, when running `make check-rust` with `RUNTESTFLAGS='--target_board=unix\{,-m32,-mx32\}'`, for example, we get `PASS` for the (default) `-m64` variant, but `FAIL: rust/debug/win64-abi.rs scan-assembler %ecx, %ecx` for the latter two variants. The `target { x86_64-*-* }` does apply in this case (and in fact should be extended/generalized to `target i?86-*-* x86_64-*-*`), but will need separate `scan-assembler` directives, conditionalized with `target [...]` or `target { ! [...] }`, with `[...]` being `lp64`/`ia32`/`x32`/etc. as appropriate (see `gcc/doc/sourcebuild.texi`, "Effective-Target Keywords".

I have tested, `win64` ABI is unsupported on 32-bit x86 targets in `rustc`. `rustc` will report:

```
error[E0570]: `"win64"` is not a supported ABI for the current target
 --> <source>:2:1
  |
2 | pub extern "win64" fn square(num: i32) -> i32 {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```